### ISSUE TITLE
Consider using the Flow ID as stream chunk ID instead of having chunk types

### ISSUE THREAD

**Issue by tfpauly** [ietf-wg-masque/draft-ietf-masque-connect-udp#38]
Currently, using CONNECT-UDP over DATAGRAMs and over HTTP streams use entirely different mechanisms for identifying data:

- DATAGRAMS use a Datagram Flow ID, which allows the request headers to define meanings and mappings
- The fallback to stream case uses an IANA registry of chunk types

I think we can simplify this by always using the Flow ID. The Flow ID can replace the "type" in the stream chunk, making the content of the chunks identical to DATAGRAM frames (length + flow ID + payload).

**Comment by mirjak** [comment]
I guess alternatively you could add the chunk type to the datagram (instead of overloading the flow ID). That consumes a few more bits but maybe also gives more flexibility.

**Comment by tfpauly** [comment]
@mirjak the way I view it, the flow ID is an arbitrary varint field, and the chunk type is too; having two arbitrary fields doesn't seem to be useful?

**Comment by mirjak** [comment]
It's two type fields in two different protocols (H3 and Masque). In that sense it's cleaner to separate them. Depends on how urgently we want to save bits.

**Comment by tfpauly** [comment]
The flow ID for H3 doesn't have meaning in and of itself beyond demulitplexing—but MASQUE can bring meaning to it.

Since the current draft presents the stream-chunk approach as a compatibility mechanism for H2 or H1, I view it as something trying to provide the semantics of H3 datagrams for older versions. To that end, my proposal is to make that very explicit and define that for H2 and H1, they can express datagram semantics over streams using the same identifiers as H3.

**Comment by DavidSchinazi** [comment]
How would this proposal work when datagrams are disabled (or with HTTP/2)?

**Comment by tfpauly** [comment]
The flow ID is completely determined by the HTTP layer, and isn't really specific to the use of datagrams--it's just an identifier for a set of messages. The fact that those messages get encoded over QUIC datagrams is a wire image detail, and it seems reasonable to map those when in cases where the wire image is over a stream.

**Comment by DavidSchinazi** [comment]
Right now we use the "Datagram-Flow-Id" header to negotiate datagram support, so we'd need a different mechanism for that. That increase in complexity might outweigh the simplicity benefits of unifying chunk IDs and flow IDs.

**Comment by tfpauly** [comment]
@DavidSchinazi Doesn't the settings parameter determine if datagrams are supported? My assumption would be that if you send the flow ID without supporting datagrams, you need to encode them in the stream. 

I really think that a separate registry of types just for the fallback case of no datagrams seems very awkward, and I don't see yet why it is necessary.

**Comment by kazuho** [comment]
To me it seems that there are two things being discussed here:
* if chunk ID of a stream chunk indicates the type of data being sent, or if it is just an identifier of a flow
* if the chunk ID field can be replaced by an identifier of a flow

To me it seems that the latter would only be sensible when we agree that the chunk ID field of a stream chunk is just an identifier, rather than something that indicates a type. Are we agreeing on that? If so, I think we can agree that there's no need for an IANA registry.

Assuming that we'd agree on that, the question becomes if calling both of the two "datagram flow IDs" makes sense. I do not have a strong opinion, though I'd point out that the scope of the identifier is going to be different for stream chunks and H3_DATAGRAM frames. For the former, the scope would be per-HTTP-stream, whereas for the latter the scope is per-HTTP-connection. Therefore, I can see some raising concern of calling things the same that would not be used interchangeably.

**Comment by DavidSchinazi** [comment]
@tfpauly that's not quite how that works in the presence of HTTP intermediaries, the SETTING only tells you if datagrams are supported in general, but you also need to know if they're supported for this specific request - as an intermediary might want to support datagrams only to certain backends. So, as @kazuho was saying, per-connection vs per-stream.

**Comment by tfpauly** [comment]
Ah, yes, quite right. The settings lets you know if the connection you're on supports datagrams.

If the intermediary did support datagrams, however, could it not translate those into stream chunks if it couldn't speak datagrams upstream (assuming we shared ID space)? If there is a deterministic translation, I could still get the benefit of datagrams between the client and intermediary, which often might be the hop where it matters most.

**Comment by DavidSchinazi** [comment]
An intermediary could definitely do that, but the current design doesn't require all intermediaries to implement this, so this would increase intermediary implementation complexity

**Comment by tfpauly** [comment]
Could we require intermediaries to do this if they support the SETTINGS option, and allow upstreams to servers that don't support datagrams?

**Comment by DavidSchinazi** [comment]
We could, but why are we doing this in the first place? I thought the argument was simplicity

**Comment by tfpauly** [comment]
I'm going for simplicity in the spec, simplicity for designing extensions, and simplicity for endpoints that are deployed in the most common cases.

Currently, the registry of chunk types is only relevant to implementations when they're falling back to cases where datagrams aren't supported. This means that all implementations—clients, servers, and intermediaries—need to maintain separate logic for how messages are mapped. Also, if I want to write an extension (let's say, to do ECN), I'll need to define how to encode the meaning of my datagrams via flow IDs _and_ entirely separately define chunk types for the meanings of different chunks.

At the end of the day, the information being communicated should be isomorphic between datagrams and stream chunks; so, it should be possible to represent the types of messages with (at least) isomorphic identifiers, if not exactly the same identifiers. It seems very odd that one is negotiated through headers dynamically and the other has an IANA registry.

Now, given that I'm not an intermediary implementor, I may not be understanding the complexity correctly, but it seems to me that the system is overall simpler and more efficient if the layer that is aware of translation already is the one that is responsible for mapping between datagrams and stream chunks, and can do it with consistent identifiers. This means a client or end server would only need to do one-or-the-other on a connection, not a mix of both.

If I'm a client, and my connection's server says it supports datagrams, I'm going to prefer to send a DATAGRAM frame right after my CONNECT-UDP request to avoid an extra round trip. If the server is an intermediary whose upstream doesn't support datagrams, I'd rather it either (a) not advertise datagram support in its own SETTINGS (which is easy for the intermediary, and means the client can start sending stream chunks right away), or (b) do the deterministic translation into stream chunks (and get the benefit of being able to use datagrams on the hop between the client and intermediary). Currently, it would (c) receive the client's datagram, but drop it on the floor, and send a response back to tell the client to send the same message again over the stream. This seems to be more complex to me, overall; and certainly less optimal.

**Comment by kazuho** [comment]
@tfpauly I would point out that there's at least one difference other than flow IDs vs. chunks, and that is negotiation. I would not be surprised if the difference becomes a headache in designing extensions.

Using ECN as an example, I'd assume that the ECN signal will be conveyed by a different flow, designated using a parameter of the datagram-flow-id header field, e.g.,`datagram-flow-id: 0; ecn=2`.

The exchange of the header field serves two purposes; 1) negotiation of the features to be used, 2) dynamic mapping between flow IDs and types.

Because use of ECN is negotiated using the header fields, we can send ECN-marked datagrams using flow ID 2, while sending non-marked datagrams using ID 0.

But in case of stream chunks, there is no negotiation. Stream chunks carrying unknown ID are simply discarded. Therefore, even if we define a stream chunk type for ECN, that type cannot carry the payload of ECN-marked datagrams. We have to define a way to split ECN signal and datagram payload, and send them separately.

To summarize, to me it seems that the debate is between:
* Proposed change) For both stream chunks and H3_DATAGRAM, use flow IDs, use dynamic mapping, with negotiation. Features are named by the header fields.
* Status quo) For stream chunks continue using static mapping without negotiation. Features are identified by stream chunk types.

Am I reading correctly?

**Comment by tfpauly** [comment]
Yes, my proposal is that the stream chunks are negotiated exactly the same as datagram flow identifiers. Effectively, they become message flow identifiers, where the messages can be sent over LV stream chunks or LV datagram frames. They could use the same headers to negotiate, and key only off of settings to know which to use for a given hop.

**Comment by kazuho** [comment]
@tfpauly Thank you for the clarification. If we are to take that route (of negotiating features), I would be supportive to adopting the proposed change. The benefit of having one negotiation mechanism outweighs the cost of having to map types dynamically when only using stream chunks.

**Comment by mirjak** [comment]
+1 to basically use the same negotiation mechanism and therefore make fallback to streams easy. I also think that using datagram where possible and potentially "translate" at the intermedia is a good thing. I think still we might have two different pieces of information here, the flow/stream ID and some kind of chunk/message header. I think keeping these things separate makes the protocol more flexible (even though that might cost a few more bits which however its probably only rally a problem for datagrams but not so much for streams).

**Comment by DavidSchinazi** [comment]
@tfpauly I don't understand how this negotiation would work. Could you please write up an example of what the HTTP headers on request and response would look like? Note that this needs to support optional negotiation of datagram support, and needs to work in the presence of intermediaries.

**Comment by tfpauly** [comment]
Yes, it's quite simple:

Server supports datagrams:
--> SETTINGS(H3_DATAGRAMS)
<-- SETTINGS(H3_DATAGRAMS)
--> HEADERS(CONNECT-UDP, datagram-flow-id: 2)
<-- HEADERS(CONNECT-UDP, datagram-flow-id: 2)
--> DATAGRAM(id=2, body)
<-- DATAGRAM(id=2, body)

Server doesn't support datagrams:
--> SETTINGS(H3_DATAGRAMS)
<-- SETTINGS()
--> HEADERS(CONNECT-UDP, datagram-flow-id: 2)
<-- HEADERS(CONNECT-UDP, datagram-flow-id: 2)
--> DATA(length=N,id=2, body)
<-- DATA(length=M,id=2, body)

And if intermediaries want to talk datagrams to clients but the upstream server doesn't, they just stuff DATAGRAMS over DATA frame chunks and vice versa.

The only thing you don't have here is the ability to have a request-by-request ability to choose if it should send CONNECT-UDP messages over DATAGRAM frames or DATA frame chunks. However, I don't see how that is important.

**Comment by tfpauly** [comment]
Okay, spoke in gather town... proposal would be to keep stream chunks, but when you have a data type of chunk, you still include the flow ID always. Make it clear that you can have other stream chunk types even when supporting datagrams as a way to support reliable control messages associated with the request.

This way the stream chunk data content is always isomorphic to datagram payloads. This does allows intermediaries to map between datagram and chunks without loss of data.

**Comment by tfpauly** [comment]
And you'd have to reset the stream if you receive a stream chunk datagram ID that isn't defined for this stream.

**Comment by kazuho** [comment]
I might want to push back a bit on going back to status quo, in particular as an argument for having a unified negotiation scheme.

I think that there might a subtle error in @tfpauly's example, but I think that the concept is visible. Use one mechanism that maps types to flow IDs, and use one flow ID in both H3_DATAGRAM frames and stream chunks to indicate the type of data being sent. We can have a separate signal (e.g., a header field) indicating if H3_DATAGRAMs can be used on a particular stream.

To give an example, if the peers ended up negotiating `datagram-flow-id: 2, 4=timestamp`, flow ID of 2 would indicate that the payload is a datagram without timestamp, and that flow ID of 4 would indicate that the payload is a datagram with a timestamp, regardless of the payload being carried by a H3_DATAGRAM frame or as a stream chunk.

I think there's appetite among us to have a negotiation mechanism that is isomorphic.

However, there's one problem, and that is the "datagram-flow-id" header field. The field is defined as a way to communicate the use of H3_DATAGRAM frame, instead of a way of communicating how the CONNECT-UDP method would convey data regardless of the conveyor being H3_DATAGRAM frame or a stream chunk.

I think that might be the crux of the problem.

Do we want have two negotiation methods for CONNECT-UDP, one that negotiates the use of H3_DATAGRAM frames using the "datagram-flow-id" header field, as well as defining a scheme to negotiate the types to be used for stream chunks? Or do we want to create an isomorphic negotiation scheme for CONNECT-UDP, that designates IDs for both H3_DATAGRAM flows and stream chunk flows?

ATM my preference goes to the latter.

Honestly, I'm not sure if I like the current approach being taken in H3_DATAGRAM, of designing `datagram-flow-id` header field as the only way to negotiate how flows are to be used. The scheme being discussed is complex, and it could be the case that the complexity roots from the fact that the scheme has to cover all use cases including the ones that we might see in the future.

We could take a different path, allowing each application of H3_DATAGRAM to designate the flow IDs that it would use in its own way. If we take that path, masque can name the flow IDs isomorphically, in a simple manner.

**Comment by LPardue** [comment]
> Honestly, I'm not sure if I like the current approach being taken in H3_DATAGRAM, of designing datagram-flow-id header field as the only way to negotiate how flows are to be used. The scheme being discussed is complex, and it could be the case that the complexity roots from the fact that the scheme has to cover all use cases including the ones that we might see in the future.

> We could take a different path, allowing each application of H3_DATAGRAM to designate the flow IDs that it would use in its own way. If we take that path, masque can name the flow IDs isomorphically, in a simple manner.

I think that's a fair observation but during the "dispatching" of the H3_DATAGRAM draft to a working group, there were some strong opinions against defining a generic HTTP/3 mechanism that was not tied to any semantic. The header is a binding. It could be substituted if someone wanted to (and they could probably gain a lot of inspiration from the text that we'll have on allocating and managing flow IDs.)

That said, I think I agree with your observation:

> The field is defined as a way to communicate the use of H3_DATAGRAM frame, instead of a way of communicating how the CONNECT-UDP method would convey data regardless of the conveyor being H3_DATAGRAM frame or a stream chunk.

There's a subtlety between these two that I think is easily missed and contributing some of the difficulty in discussion.

**Comment by DavidSchinazi** [comment]
Overcome by events: flow IDs were removed in 649890881aaa92548b5a0fb37a7622b409b281fd. Closing.
